include/pybind11/pybind11.h

```diff
@@ -223,20 +231,48 @@ class cpp_function : public function {
         }
     }
 
+    // Utility class that keeps track of all duplicated strings, and cleans them up in its destructor,
+    // unless they are released. Basically a RAII-solution to deal with exceptions along the way.
+    class strdup_guard {
+    public:
+        ~strdup_guard() {
+            for (auto s : strings)
+                std::free(s);
+        }
+        char *operator()(const char *s) {
+            auto t = strdup(s);
+            strings.emplace_back(t);
+            return t;
+        }
+        void release() {
+            strings.clear();
+        }
+    private:
+        std::vector<char *> strings;
+    };
+
     /// Register a function call with Python (generic non-templated code goes here)
-    void initialize_generic(detail::function_record *rec, const char *text,
+    void initialize_generic(unique_function_record &&unique_rec, const char *text,
```

Comment:
Why are you taking `std::unique_ptr` by reference instead of by-value? Remember that r-value references mean "maybe move".

Considering that we don't really want the unique_ptr here, rather just the pointer and a way to tell the unique_ptr it isn't in charge, maybe take by const-ref? A const-ref-unique-pointer has weird semantic meaning, transferring ownership from unique pointer to a capsule is probably equally weird.

Reply:
> Considering that we don't really want the unique_ptr here, rather just the pointer and a way to tell the unique_ptr it isn't in charge, maybe take by const-ref?

That not going to work, because it needs to give up ownership to the `capsule`, later. I really want to pass ownership, not just `initialize_generic` use it.

> Remember that r-value references mean "maybe move".

True. So, ordinarily by value?

Reply:
Combined with https://github.com/pybind/pybind11/pull/2756#discussion_r551519918, this is even worse, though, the accept by value :-/

At least with `unique_function_record &&`, you have the contract "we only move in case of success, and otherwise the caller keeps ownership"

Reply:
> > Considering that we don't really want the unique_ptr here, rather just the pointer and a way to tell the unique_ptr it isn't in charge, maybe take by const-ref?
> 
> That not going to work, because it needs to give up ownership to the `capsule`, later. I really want to pass ownership, not just `initialize_generic` use it.
> 

Right, but passing by mutable reference would work. Though you know what I think about mutable references.

> > Remember that r-value references mean "maybe move".
> 
> True. So, ordinarily by value?

I guess so.

Reply:
> Right, but passing by mutable reference would work. Though you know what I think about mutable references.

Is that any cleaner? At least now we know that we're "maybe moving"?

If you have a nice combo with https://github.com/pybind/pybind11/pull/2756#discussion_r551519918, I'll implemented it (or feel free to push something), but otherwise...

Reply:
> a nice combo with #2756 (comment)

Unfortunately not. Let's just go with pass by value.

Reply:
Added a comment on why we should probably not do so: https://github.com/pybind/pybind11/pull/2756/commits/9942614495cada974791019b1644d9bb07d21a8d